src/qtism/data/storage/xml/versions/QtiVersion.php

```diff
@@ -0,0 +1,171 @@
+    {
+        return $this->versionNumber;
+    }
+
+    /**
+     * Creates a new Version given the version number.
+     *
+     * @param string $versionNumber
+     * @return $this
+     */
+    public static function create(string $versionNumber): self
+    {
+        $versionNumber = self::sanitize($versionNumber);
+        $class = static::SUPPORTED_VERSIONS[$versionNumber];
+        return new $class($versionNumber);
+    }
+
+    /**
+     * Checks that the given version is supported.
+     *
+     * @param string $version a semantic version
+     * @throws InvalidArgumentException when the version is not supported.
+     */
+    protected static function checkVersion(string $version)
+    {
+        if (!isset(static::SUPPORTED_VERSIONS[$version])) {
+            throw QtiVersionException::unsupportedVersion(static::UNSUPPORTED_VERSION_MESSAGE, $version, static::SUPPORTED_VERSIONS);
+        }
+    }
+
+    /**
+     * Infer the QTI version of the document from its XML definition.
+     *
+     * @param DOMDocument $document
+     * @return string a semantic QTI version inferred from the document.
+     * @throws XmlStorageException when the version can not be inferred.
+     */
+    public static function infer(DOMDocument $document): string
+    {
+        $root = $document->documentElement;
+        $version = '';
+
+        if ($root !== null) {
+            $rootNs = $root->namespaceURI;
+            if ($rootNs !== null) {
+                $version = static::findVersionInDocument($rootNs, $document);
+            }
+        }
+
+        if ($version === '') {
+            $msg = 'Cannot infer QTI version. Check namespaces and schema locations in XML file.';
+            throw new XmlStorageException($msg, XmlStorageException::VERSION);
+        }
+
+        return $version;
+    }
+
+    /**
+     * Finds the version of the document given the namespace and Xsd location.
+     *
+     * @param string $rootNs
+     * @param DOMDocument $document
+     * @return string
+     */
+    public static function findVersionInDocument(string $rootNs, DOMDocument $document): string
+    {
+        $version = '';
+
+        if ($rootNs === QtiVersion200::XMLNS) {
+            $version = '2.0.0';
+        } elseif ($rootNs === QtiVersion210::XMLNS) {
+            $version = '2.1.0';
+
+            $nsLocation = Utils::getXsdLocation($document, $rootNs);
+            if ($nsLocation === QtiVersion211::XSD) {
+                $version = '2.1.1';
+            }
+        } elseif ($rootNs === QtiVersion220::XMLNS) {
+            $version = '2.2.0';
+
+            $nsLocation = Utils::getXsdLocation($document, $rootNs);
+            if ($nsLocation === QtiVersion221::XSD) {
+                $version = '2.2.1';
+            } elseif ($nsLocation === QtiVersion222::XSD) {
+                $version = '2.2.2';
+            }
+        } elseif ($rootNs === QtiVersion300::XMLNS) {
+            $version = '3.0.0';
+        }
+
+        return $version;
```

Comment:
Here I would also suggest using `switch` and early (as soon NS matched) return to improve readability

Reply:
To something like 
```
switch ($rootNs) {
    case QtiVersion200::XMLNS:
        return '2.0.0';
    case QtiVersion210::XMLNS:
        if ($nsLocation === QtiVersion211::XSD) {
            return '2.1.1';
        }

        return '2.1.0';
    case QtiVersion220::XMLNS:
        if ($nsLocation === QtiVersion221::XSD) {
            return '2.2.1';
        }

        if ($nsLocation === QtiVersion222::XSD) {
            return '2.2.2';
        }

        return '2.2.0';
    case QtiVersion300::XMLNS:
        return '3.0.0';
    default:
        return '';
}
```

Reply:
Correct. As said in other comment, I wanted to do something with the SUPPORTED_VERSIONS constant, but didn't come to something very obvious. Will change that to switch.

Reply:
I'm still against switch statements as stated here - https://github.com/oat-sa/qti-sdk/pull/201#discussion_r470626536